openshift_performance/ose3_perf/scripts/build_test_1.py

```diff
@@ -0,0 +1,414 @@
+    #    "--token=yPeDkR0b0ESwbX2WlC4AGWqQIHJHIeRbrEOOua9HEoQ")
+
+
+def run_build(build_def):
+    build_regex = re.compile("build \"(.*)\" started")
+
+    namespace = build_def["namespace"]
+    name = build_def["name"]
+    build_name = "UNKNOWN"
+
+    try:
+
+        build_name = run("oc start-build -n " + namespace + " " + name).rstrip()
+        # 3.2 Comment out following line
+        build_name = build_regex.search(build_name).group(1)
+        global_build_status[namespace + ":" + build_name] = STATUS_STARTED
+        logger.info("Build is: " + namespace + ":" + build_name)
+
+    except subprocess.CalledProcessError as err:
+        if "UNKNOWN" != build_name:
+            global_build_status[namespace + ":" + build_name] = STATUS_ERROR
+        logger.error("Command failed:  tproject=" + namespace + ",cmd=" +
+                     err.cmd + ", retcode=" + str(err.returncode) + ", output="
+                     + err.output)
+    except Exception as e:
+        logger.error("cannot get build_name from: " + build_name)
+        logger.error(e)
+
+
+def check_build_status(executor):
+    logger.info("check_build_status ...")
+    # wait the first build is started
+    time.sleep(20)
+    while not all_builds_completed():
+        try:
+            time.sleep(10)
+            result = run("oc get build --all-namespaces --no-headers")
+            parse(executor, result)
+        except Exception as e:
+            logger.error(e)
+
+
+def all_builds_completed():
+    logger.debug("global_build_status: " + str(global_build_status))
+    for key, value in global_build_status.iteritems():
+        if value < 300:
+            return False
+    return True
+
+
+def parse(executor, result):
+    for line in result.splitlines():
+        words = line.split()
+
+        if len(words) >= 5:
+            status = words[4]
+            namespace = words[0]
+            name = words[1]
+            idx = namespace + ":" + name
+            if (status.startswith("Failed")) or (status == "Cancelled") or \
+                    (status.startswith("Error")):
+                if idx in global_build_status.keys():
+                    logger.debug(idx + " FAILED")
+                    if global_build_status[idx] < STATUS_NOT_COMPLETE:
+                        global_build_status[idx] = STATUS_NOT_COMPLETE
+                        stats_idx = idx[0:idx.rindex('-')]
+                        global_build_stats[stats_idx]["failed"] += 1
+            elif "Complete" == words[4]:
+                if idx in global_build_status.keys():
+                    if global_build_status[idx] < STATUS_COMPLETE:
+                        logger.info(idx + " Complete")
+                        global_build_status[idx] = STATUS_COMPLETE
+                    duration_string = words[-1]
+                    if global_build_status[idx] < STATUS_LOGGING:
+                        executor.submit(do_post_actions, namespace, name,
```

Comment:
Is there a big benefit to having each of the do_post_actions run on its own thread vs. just calling it in-line?   More for my curiosity than any specific concern.

Reply:
To me, slow things are planned to be done in another thread unless it could cause trouble.
Here it is considered slow because it executes oc-command. :)

Reply:
Yep, you're right - pulling each build's logs in do_post_actions.